Source/WebCore/css/parser/CSSPropertyParser.cpp

```diff
@@ -5536,44 +5532,32 @@ bool CSSPropertyParser::consumeColumns(bool important)
 {
     RefPtr<CSSValue> columnWidth;
     RefPtr<CSSValue> columnCount;
-    bool hasPendingExplicitAuto = false;
-    
+
     for (unsigned propertiesParsed = 0; propertiesParsed < 2 && !m_range.atEnd(); ++propertiesParsed) {
-        if (!propertiesParsed && m_range.peek().id() == CSSValueAuto) {
+        if (m_range.peek().id() == CSSValueAuto) {
             // 'auto' is a valid value for any of the two longhands, and at this point
             // we don't know which one(s) it is meant for. We need to see if there are other values first.
             consumeIdent(m_range);
-            hasPendingExplicitAuto = true;
         } else {
-            if (!columnWidth) {
-                if ((columnWidth = consumeColumnWidth(m_range)))
-                    continue;
-            }
-            if (!columnCount) {
-                if ((columnCount = consumeColumnCount(m_range)))
-                    continue;
-            }
+            if (!columnWidth && (columnWidth = consumeColumnWidth(m_range)))
+                continue;
+            if (!columnCount && (columnCount = consumeColumnCount(m_range)))
+                continue;
             // If we didn't find at least one match, this is an invalid shorthand and we have to ignore it.
             return false;
         }
     }
-    
+
     if (!m_range.atEnd())
         return false;
 
-    // Any unassigned property at this point will become implicit 'auto'.
-    if (columnWidth)
-        addProperty(CSSPropertyColumnWidth, CSSPropertyInvalid, columnWidth.releaseNonNull(), important);
-    else {
-        addProperty(CSSPropertyColumnWidth, CSSPropertyInvalid, CSSValuePool::singleton().createIdentifierValue(CSSValueAuto), important, !hasPendingExplicitAuto /* implicit */);
-        hasPendingExplicitAuto = false;
-    }
-    
-    if (columnCount)
-        addProperty(CSSPropertyColumnCount, CSSPropertyInvalid, columnCount.releaseNonNull(), important);
-    else
-        addProperty(CSSPropertyColumnCount, CSSPropertyInvalid, CSSValuePool::singleton().createIdentifierValue(CSSValueAuto), important, !hasPendingExplicitAuto /* implicit */);
-    
+    // If both values are auto, set column-width explicitly to auto so the shorthand serializes correctly.
+    if (!columnWidth && !columnCount)
+        columnWidth = CSSValuePool::singleton().createIdentifierValue(CSSValueAuto);
```

Comment:
This behavior affects the columns shorthand, but also what the `column-width` and `column-count` properties return. Can we add a test that sets the `columns` to `auto` and then verifies how each of `column-count` and `column-width` serialize in that case? Is it correct that `column-width` serializes as `initial`?

Reply:
This is what the failing subtests are about, but I'll add a more explicit one.